### PR TITLE
[WIP] ClusterClass support in CAPX

### PR DESCRIPTION
**Pull request by deepakm-ntnx**
Ref: https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/pull/333
https://kubebuilder.io/plugins/go-v4-plugin

**What this PR does / why we need it**:
<pre>
 deepak.muley@C02ZG5BMLVDL  ~/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix   clusterclassv4  kubebuilder create api --group infrastructure --version v1beta1 --kind NutanixClusterTemplate
INFO Create Resource [y/n]                        
y
INFO Create Controller [y/n]                      
y
INFO Writing kustomize manifests for you to edit... 
INFO Writing scaffold for you to edit...          
INFO api/infrastructure/v1beta1/nutanixclustertemplate_types.go 
INFO api/infrastructure/v1beta1/groupversion_info.go 
INFO internal/controller/infrastructure/suite_test.go 
INFO internal/controller/infrastructure/nutanixclustertemplate_controller.go 
INFO Update dependencies:
$ go mod tidy           
INFO Running make:
$ make generate                
GOBIN=/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin /Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/scripts/go_install.sh k8s.io/code-generator/cmd/conversion-gen conversion-gen v0.23.6
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin/controller-gen-v0.8.0 object:headerFile="hack/boilerplate.go.txt" paths="./..."
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin/conversion-gen \
	--input-dirs=./api/infrastructure/v1alpha4 \
	--input-dirs=./api/infrastructure/v1beta1 \
	--build-tag=ignore_autogenerated_core \
	--output-file-base=zz_generated.conversion \
	--go-header-file=./hack/boilerplate.go.txt
Next: implement your new API and generate the manifests (e.g. CRDs,CRs) with:
$ make manifests
</pre>

<pre>
 kubebuilder create webhook --group  infrastructure --version v1beta1 --kind NutanixClusterTemplate --defaulting --programmatic-validation --force
INFO Writing kustomize manifests for you to edit... 
ERRO Unable to find the target #- path: manager_webhook_patch.yaml to uncomment in the file config/default/kustomization.yaml. 
INFO Writing scaffold for you to edit...          
INFO api/infrastructure/v1beta1/nutanixclustertemplate_webhook.go 
INFO api/infrastructure/v1beta1/webhook_suite_test.go 
INFO Update dependencies:
$ go mod tidy           
go: finding module for package sigs.k8s.io/controller-runtime/pkg/metrics/server
go: found sigs.k8s.io/controller-runtime/pkg/metrics/server in sigs.k8s.io/controller-runtime v0.16.3
go: finding module for package gotest.tools/v3/skip
go: finding module for package gotest.tools/v3/assert
go: finding module for package gotest.tools/v3/env
go: finding module for package gotest.tools/v3/assert/cmp
go: found gotest.tools/v3/assert in gotest.tools/v3 v3.5.1
go: found gotest.tools/v3/assert/cmp in gotest.tools/v3 v3.5.1
go: found gotest.tools/v3/env in gotest.tools/v3 v3.5.1
go: found gotest.tools/v3/skip in gotest.tools/v3 v3.5.1
INFO Running make:
$ make generate                
GOBIN=/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin /Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/scripts/go_install.sh k8s.io/code-generator/cmd/conversion-gen conversion-gen v0.23.6
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin/controller-gen-v0.8.0 object:headerFile="hack/boilerplate.go.txt" paths="./..."
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin/conversion-gen \
	--input-dirs=./api/infrastructure/v1alpha4 \
	--input-dirs=./api/infrastructure/v1beta1 \
	--build-tag=ignore_autogenerated_core \
	--output-file-base=zz_generated.conversion \
	--go-header-file=./hack/boilerplate.go.txt
Next: implement your new Webhook and generate the manifests with:
$ make manifests
</pre>

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**How Has This Been Tested?**:
<pre>
 ✘ deepak.muley@C02ZG5BMLVDL  ~/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix   clusterclassv4  make manifests
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin/controller-gen-v0.8.0 "crd:crdVersions=v1" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
 deepak.muley@C02ZG5BMLVDL  ~/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix   clusterclassv4  make cluster-templates
GOBIN=/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin /Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/scripts/go_install.sh sigs.k8s.io/kustomize/kustomize/v4 kustomize v4.5.4
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin/kustomize-v4.5.4 build templates/base > templates/cluster-template.yaml
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin/kustomize-v4.5.4 build templates/csi > templates/cluster-template-csi.yaml
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin/kustomize-v4.5.4 build templates/ccm > templates/cluster-template-ccm.yaml
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin/kustomize-v4.5.4 build templates/clusterclass > templates/cluster-template-clusterclass.yaml
 deepak.muley@C02ZG5BMLVDL  ~/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix   clusterclassv4  make all
GOBIN=/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin /Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/scripts/go_install.sh k8s.io/code-generator/cmd/conversion-gen conversion-gen v0.23.6
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin/controller-gen-v0.8.0 object:headerFile="hack/boilerplate.go.txt" paths="./..."
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin/conversion-gen \
	--input-dirs=./api/infrastructure/v1alpha4 \
	--input-dirs=./api/infrastructure/v1beta1 \
	--build-tag=ignore_autogenerated_core \
	--output-file-base=zz_generated.conversion \
	--go-header-file=./hack/boilerplate.go.txt
go fmt ./...
go vet ./...
go build -o bin/manager cmd/main.go
</pre>

<pre>
 deepak.muley@C02ZG5BMLVDL  ~/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix   clusterclassv4  clusterctl generate cluster ccls-test1 --from ./templates/cluster-template-clusterclass.yaml -n workloads > ccls-test1.yaml
Using configuration File="/Users/deepak.muley/.cluster-api/clusterctl.yaml"
Using configuration File="/Users/deepak.muley/.cluster-api/clusterctl.yaml"
</pre>

<pre>
kustomize edit fix
# Warning: 'patchesStrategicMerge' is deprecated. Please use 'patches' instead. Run 'kustomize edit fix' to update your Kustomization automatically.

Fixed fields:
  patchesJson6902 -> patches
  patchesStrategicMerge -> patches
  commonLabels -> labels

To convert vars -> replacements, run the command `kustomize edit fix --vars`

WARNING: Converting vars to replacements will potentially overwrite many resource files 
and the resulting files may not produce the same output when `kustomize build` is run. 
We recommend doing this in a clean git repository where the change is easy to undo.
</pre>

<pre>
 deepak.muley@C02ZG5BMLVDL  ~/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix   clusterclassv4 ±  make cluster-templates
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin/kustomize-v5.2.1 build templates/base > templates/cluster-template.yaml
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin/kustomize-v5.2.1 build templates/csi > templates/cluster-template-csi.yaml
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin/kustomize-v5.2.1 build templates/ccm > templates/cluster-template-ccm.yaml
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin/kustomize-v5.2.1 build templates/clusterclass > templates/cluster-template-clusterclass.yaml
</pre>

_Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and test output_
<pre>
 deepak.muley@C02ZG5BMLVDL  ~/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix   clusterclassv4 ±  make test-clusterclass-create
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin/kustomize-v5.2.1 build templates/base > templates/cluster-template.yaml
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin/kustomize-v5.2.1 build templates/csi > templates/cluster-template-csi.yaml
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin/kustomize-v5.2.1 build templates/ccm > templates/cluster-template-ccm.yaml
/Users/deepak.muley/go/src/github.com/deepakm-ntnx/cluster-api-provider-nutanix/hack/tools/bin/kustomize-v5.2.1 build templates/clusterclass > templates/cluster-template-clusterclass.yaml
clusterctl generate cluster ccls-test1 --from ./templates/cluster-template-clusterclass.yaml -n capx-test-ns > ccls-test1.yaml
Using configuration File="/Users/deepak.muley/.cluster-api/clusterctl.yaml"
Using configuration File="/Users/deepak.muley/.cluster-api/clusterctl.yaml"
kubectl create ns capx-test-ns || true
Error from server (AlreadyExists): namespaces "capx-test-ns" already exists
kubectl apply -f ./ccls-test1.yaml
configmap/user-ca-bundle created
secret/ccls-test1 configured
kubeadmconfigtemplate.bootstrap.cluster.x-k8s.io/my-test-cluster-template-md-kcfgt created
Warning: Cluster refers to ClusterClass my-test-cluster-template in the topology but it does not exist. Cluster topology has not been fully validated. The ClusterClass must be created to reconcile the Cluster
cluster.cluster.x-k8s.io/ccls-test1 created
clusterclass.cluster.x-k8s.io/my-test-cluster-template created
kubeadmcontrolplanetemplate.controlplane.cluster.x-k8s.io/my-test-cluster-template-kcpt created
nutanixclustertemplate.infrastructure.cluster.x-k8s.io/my-test-cluster-template-nct created
nutanixmachinetemplate.infrastructure.cluster.x-k8s.io/my-test-cluster-template-cp-nmt created
nutanixmachinetemplate.infrastructure.cluster.x-k8s.io/my-test-cluster-template-md-nmt created
</pre>

<pre>
first nutanix clusterclass based cluster is getting created now
NAME                                                     AGE
clusterclass.cluster.x-k8s.io/my-test-cluster-template   3m55s
NAME                                  CLUSTERCLASS               PHASE         AGE     VERSION
cluster.cluster.x-k8s.io/ccls-test1   my-test-cluster-template   Provisioned   2m45s   v1.28.4
NAME                                                                         ADDRESS        READY   PROVIDERID                                       NODEREF
nutanixmachine.infrastructure.cluster.x-k8s.io/ccls-test1-md-0-9lpt8-5dztl   10.45.63.87    true    nutanix://41d77d9a-6052-49df-bc74-5500620fdcd4   ccls-test1-md-0-tfftv-sr46w-nf7hb
nutanixmachine.infrastructure.cluster.x-k8s.io/ccls-test1-md-0-9lpt8-f65j6   10.45.63.110   true    nutanix://aa7b4b68-adde-46e8-b25d-40305fcf419d   
nutanixmachine.infrastructure.cluster.x-k8s.io/ccls-test1-zzmcp-z2qzm        10.45.63.77    true    nutanix://264e38d1-30fb-4be3-bca7-30a39d980561   ccls-test1-cck8w-t6w2m
</pre>

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by jimmidyson** [comment]
@deepakm-ntnx I'm going to look at this on Monday, but is there any chance you could split this PR up into smaller, logical chunks? It would make reviewing much easier, especially coming at it with little context like me 😅 I was thinking it could be broken down into: kustomization edit fixes; kubebuilder plugin v4 layout changes (including moving api package); adding ClusterClass template API kinds. WDYT?

**Comment by deepakm-ntnx** [comment]
> @deepakm-ntnx I'm going to look at this on Monday, but is there any chance you could split this PR up into smaller, logical chunks? It would make reviewing much easier, especially coming at it with little context like me 😅 I was thinking it could be broken down into: kustomization edit fixes; kubebuilder plugin v4 layout changes (including moving api package); adding ClusterClass template API kinds. WDYT?

Sure, but its still WIP. will do that once tests are working. wasn't planning to add reviewers yet

File: Makefile [hunk 1]
```diff
@@ -235,25 +235,25 @@ kind-delete: ## Delete the kind cluster
...
 docker-push: $(KO) ## Push docker image with the manager.
-	KO_DOCKER_REPO=${IMG_REPO} $(KO) build --bare --platform=${PLATFORMS} -t ${IMG_TAG} .
+	KO_DOCKER_REPO=${IMG_REPO} $(KO) build --bare --platform=${PLATFORMS} -t ${IMG_TAG} ./cmd
 
 .PHONY: docker-push-kind
 docker-push-kind: $(KO) ## Make docker image available to kind cluster.
-	GOOS=linux GOARCH=${shell go env GOARCH} KO_DOCKER_REPO=ko.local ${KO} build -B -t ${IMG_TAG} -L .
-	docker tag ko.local/cluster-api-provider-nutanix:${IMG_TAG} ${IMG}
+	GOOS=linux GOARCH=${shell go env GOARCH} KO_DOCKER_REPO=ko.local ${KO} build -B -t ${IMG_TAG} -L ./cmd
+	docker tag ko.local/cmd:${IMG_TAG} ${IMG}
```
Comment (faiq) [inline Makefile:256]: copy paste mistake.

File: Makefile [hunk 2]
```diff
@@ -328,14 +328,15 @@ cluster-templates: $(KUSTOMIZE) ## Generate cluster templates for all flavors
...
 
 ##@ Testing
 
 .PHONY: docker-build-e2e
 docker-build-e2e: $(KO) ## Build docker image with the manager with e2e tag.
 	GIT_COMMIT_HASH=`git rev-parse HEAD` && \
-	KO_DOCKER_REPO=ko.local GOFLAGS="-ldflags=-X=main.gitCommitHash=$${GIT_COMMIT_HASH}" $(KO) build -B --platform=${PLATFORMS_E2E} -t e2e -L .
-	docker tag ko.local/cluster-api-provider-nutanix:e2e ${IMG_REPO}:e2e
+	KO_DOCKER_REPO=ko.local GOFLAGS="-ldflags=-X=main.gitCommitHash=$${GIT_COMMIT_HASH}" $(KO) build -B --platform=${PLATFORMS_E2E} -t e2e -L ./cmd
+	docker tag ko.local/cmd:e2e ${IMG_REPO}:e2e
```
Comment (faiq) [inline Makefile:339]: copy paste mistake

**Pull request closed without merging**